drivers/sms_ser.c

```diff
@@ -0,0 +1,698 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Reference of the derivative work: riello driver
+ */
+
+#include "config.h" /* must be the first header */
+#include <string.h>
+
+#include "common.h" /* for upsdebugx() etc */
+#include "sms_ser.h"
+#include "main.h"
+#include "serial.h"
+
+#define ENDCHAR '\r'
+
+#define DRIVER_NAME "SMS Brazil UPS driver"
+#define DRIVER_VERSION "1.00"
+
+#define QUERY_SIZE 7
+#define BUFFER_SIZE 18
+#define RESULT_SIZE 18
+#define HUMAN_VALUES 7
+
+static unsigned int bootdelay = DEFAULT_BOOTDELAY;
+static uint8_t bufOut[BUFFER_SIZE];
+static uint8_t bufIn[BUFFER_SIZE];
+static SmsData DeviceData;
+
+/* driver description structure */
+upsdrv_info_t upsdrv_info = {
+    DRIVER_NAME,
+    DRIVER_VERSION,
+    "Alex W. Baulé <alexwbaule@gmail.com>",
+    DRV_BETA,
+    {NULL}};
+
+void sms_parse_features(uint8_t *rawvalues, SmsData *results) {
+    char tbattery[6];
+    char frequency[4];
+
+    memset(results->voltageRange, 0, sizeof(results->voltageRange));
+    memset(results->currentRange, 0, sizeof(results->currentRange));
+    memset(tbattery, 0, sizeof(tbattery));
+    memset(frequency, 0, sizeof(frequency));
+
+    for (int i = 1; i < BUFFER_SIZE - 2; i++) {
+        if (i <= 7) {
+            snprintfcat(results->voltageRange, 14, "%c", rawvalues[i]);
+        } else if (i <= 10) {
+            snprintfcat(results->currentRange, 6, "%c", rawvalues[i]);
+        } else if (i <= 13) {
+            snprintfcat(tbattery, sizeof(tbattery), "%c", rawvalues[i]);
+        } else {
+            snprintfcat(frequency, sizeof(frequency), "%c", rawvalues[i]);
+        }
+    }
+
+    results->voltageBattery = atoi(tbattery);
+    results->frequency = atoi(frequency);
+}
+
+void sms_parse_information(uint8_t *rawvalues, SmsData *results) {
+    /* Count from 1 to ignore first char and remove 2 from BUFFER_SIZE
+     *  to compensate the start and ignore '\r' from end. */
+    memset(results->model, 0, sizeof(results->model));
+    memset(results->version, 0, sizeof(results->version));
+
+    for (int i = 1; i < BUFFER_SIZE - 2; i++) {
+        if (i <= 12) {
+            snprintfcat(results->model, 24, "%c", rawvalues[i]);
+        } else {
+            snprintfcat(results->version, 6, "%c", rawvalues[i]);
+        }
+    }
+}
+
+void sms_parse_results(uint8_t *rawvalues, SmsData *results) {
+    char buf[BUFFER_SIZE];
+    unsigned char byte, mask, p;
```

Comment:
Wondering if `uint8_t` would be better than `unsigned char` here, with bitmask maths involved... just to be sure on different platforms.

Reply:
For me its ok, i only has Linux and freebsd to test it.